Fix gradcheck for CSR and CSC inputs.

pearu opened the pull request:
Partially fix-es https://github.com/pytorch/pytorch/issues/87085

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89786



cc @nikitaved @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

pearu committed:
Fix gradcheck for CSR and CSC inputs.

[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -1167,6 +1188,18 @@ def _vec_from_tensor(x, generator, downcast_complex=False):
+            .to(dtype=dtype, device=x.device) \
+            .view(x_values.shape)
+        values /= values.norm()
+        vec = torch.sparse_compressed_tensor(compressed_indices, plain_indices, values, x.size(), layout=x.layout)
```
FYI this path is only used for fast gradcheck, so looks like it is currently not tested?

pearu replied:
The corresponding tests in `test_gradcheck_sparse_csr/csc_input` are disabled due to other issues that are OT for this PR.

pearu committed:
Update on "Fix gradcheck for CSR and CSC inputs."

Partially fix-es https://github.com/pytorch/pytorch/issues/87085




cc nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -91,8 +102,17 @@ def get_stride(size):
+            x_indices = torch._convert_indices_from_csr_to_coo(x_tensor.ccol_indices(), x_tensor.row_indices(), transpose=True).t()
+            x_values = x_tensor.values()
+        else:
+            raise NotImplementedError('_iter_tensor for {x_tensor.layout} input')
```
missing f

pearu replied:
Fixed.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -20,6 +20,17 @@ class GradcheckError(RuntimeError):
+
+
+def _is_sparse_any_tensor(obj):
+    return is_tensor_like(obj) and obj.layout in {torch.sparse_coo, torch.sparse_csr, torch.sparse_csc,
```
nit: most of this list can be deduplicated with the one above.

pearu replied:
Fixed. Also, the call to `is_tensor_like` turns out to be unnecessary because these functions are used in the context where the inputs are always tensors.

pearu committed:
Update on "Fix gradcheck for CSR and CSC inputs."

Partially fix-es https://github.com/pytorch/pytorch/issues/87085




cc nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD reviewed (approved):
SGTM thanks!

pearu commented:
@pytorchbot merge
